### PR TITLE
issue 83 logs and analysis using docker

### PR DESCRIPTION
**Pull request by abessiari**
- Docker file using ubuntu 20.04 and python 3.8.5
- Support for s3 buckets. 
- See README under docker directory. 
- Minimal modification to ansible scripts to support ansible 2.9.6
- Tested using:  sudo make all test; sudo make safe-update
- Tested Log rotation using: sudo :logrotate -v -f /etc/logrotate.d/apache2
- Added  github workflow build_and_test.yml

### CODE REVIEW

**Comment by abessiari** [comment]
@jamesaoverton 
cc: @kltm 

Hi James this pull request supersedes PR 747.
Docker support has been added. 

The only thing that I could not test is make safe-update as it keeps saying that the last build failed and thus no updates.

It would be nice if you can take a look and help test/make sure we are on the right path.
Thanks.
aes

**Comment by jamesaoverton** [comment]
Thanks! Yes, safe-update is completely broken because we ran out of free Travis credits.

I will look at this and test is as soon as I can, which unfortunately won't be for a few days.

**Comment by kltm** [comment]
Linking https://github.com/OBOFoundry/purl.obolibrary.org/issues/754

File: aws/vars.tf [hunk 1]
```diff
@@ -0,0 +1,32 @@
+variable "tags" {
+  type = map
+  default = { Name = "test-purl-obolibrary" }
+}
+
+variable "instance_type" {
+  default = "t2.large"
```
Comment (jamesaoverton) [inline aws/vars.tf:7]: We've been getting by with `t2.micro` with no problems.

File: aws/vars.tf [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+variable "ssh_port" {
+  type        = number
+  default     = 22
+  description = "ssh server port"
+}
+
+variable "http_port" {
+  type        = number
+  default     = 80
+  description = "noctua server port"
```
Comment (jamesaoverton) [inline aws/vars.tf:31]: There shouldn't be any references to Noctua.

File: aws/ssh_key.tf [hunk 3]
```diff
@@ -0,0 +1,5 @@
+resource "aws_key_pair" "ssh_key" {
+  key_name   = var.key_name
+  public_key = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC1TL5/v2/fy0kyX68/W8YT2bML7XvK6IF84vgSYDCqk2E/UbsLXYbUEk5/nxFOBW+4gEmQP9D80jkAAOEN4zXalctKokib8SMHhCSzv+epCtqsbxKzU71Gfl9elJWNWZcwnCzwkZrO6RDZeu7Jcp3Asx4NiyAKd7PjWuLU8qmJo5rjwbAS4dumvlDLdjVWka4jRT02jVP5+p2J/+JwvZJCGMlEb6d4WKuy8UsO4wZXXzgwrYRPfBu3XaZFA19MoLAO1UyzZwKSsin7yg/+JN8IE7sIwvvL9pExD4DU0krfNyjAO8c88UhzrUbN/vEGxDXEIvvmIQ8k1MtHh8o6WwP3 AEssiari@aessiari-mba"
```
Comment (jamesaoverton) [inline aws/ssh_key.tf:3]: Who should have keys here?

File: dependencies/bare-metal.yml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+name: test_purl_obolibrary
+channels:
+  - defaults
+  - conda-forge
+  - anaconda
```
Comment (jamesaoverton) [inline dependencies/bare-metal.yml:5]: Why do we need Anaconda? Why use Python 2? The PURL system requires Python 3: https://github.com/OBOFoundry/purl.obolibrary.org/blob/master/Makefile#L16

**Comment by jamesaoverton** [comment]
Sorry for the delay in replying. There's a major grant deadline this week that has taken up all of my time.

I was able to build the Docker image and run `make` inside it. I didn't try the logrotate stuff.

I guess I have a bunch of questions about the deployment story. I don't understand why all the `aws/` files are here.

File: dependencies/bare-metal.yml (hunk 4 above)
Comment (jamesaoverton) [reply dependencies/bare-metal.yml:5]: I guess this is for local testing on macOS?

**Comment by jamesaoverton** [comment]
It's really important to have CI on this repo. We ran out of Travis credits. The plan has been to use GitHub Actions but I though we'd wait for this. What is the plan for using this work with GitHub Actions?

**Comment by abessiari** [comment]
@jamesaoverton 

Thanks for looking into this.  
The aws stuff was a copy and paste from noctua. I will clean it up. 
I will take a look at the conda file as well.

**Comment by abessiari** [comment]
@jamesaoverton 
I will just remove the aws files and conda. Initially I was testing on aws. So I just thought to squezze them in. :-)

**Comment by abessiari** [comment]
@jamesaoverton 

I have removed the aws and the conda files. They are not needed at this point. 
For Github Actions please advise (some details) on what needs to happen if you want me to take a crack at it.

**Comment by jamesaoverton** [comment]
Thanks for those changes @abessiari. I just haven't been able to find a couple of uninterrupted hours to work on this. I'll keep trying.

**Comment by abessiari** [comment]
@jamesaoverton 
@kltm 

Hi. 

I added a build_and_test.yml github workflow.  
This builds the docker image and runs the test. 
It is under .github/workflows. 

Next steps is to use this for safe_update. Can you and I have a zoom meeting 
to discuss how to go about this. 
Thanks.
aes

**Comment by jamesaoverton** [comment]
Thanks @abessiari, much appreciated. I'm happy to talk with you on Zoom. Let's arrange that on the OBO community Slack, if you're on it, or just email me: james at overton.ca.

The current approach to "safe-update" is a "pull" approach for updating the PURLs. We use `cron` every 10 minutes to check for an update that has passed the CI tests. We could use a "push" approach instead, with a webhook or something. Or if we're packaging and deploying via Docker then I guess we want another approach entirely? @kltm?


If we do want to update the current "safe-update" approach, we need to make some changes. Basically tools/safe-update.py script queries the Travis CI API to see whether the master branch CI tests have succeeded. If they did succeed, we update the PURLs. If not, we don't. We aren't using Travis any more, but we still want to check the status of the GitHub Actions for the master branch.

This should be simple. For a given check like <https://github.com/obi-ontology/obi/actions/workflows/obi-test.yml> there's an easy way to get an SVG status badge. But I can't figure out a similarly easy way to get a pass/fail exit status.

This is ugly but it returns "success" for the OBI repository's master branch:

```sh
gh api repos/obi-ontology/obi/commits/$(gh api repos/obi-ontology/obi/branches/master | jq -r .commit.sha)/check-runs | jq .check_runs[0].conclusion
```

The [GitHub CLI](https://cli.github.com/) should be able to just do this, but I can't figure it out. Does anybody have a better approach?

**Comment by kltm** [comment]
Overall, I'd suggest that we treat the these as two separate problems: 1) getting everything dockerized and turnkey so that anybody with enough permissions in AWS (EC2 and Route 53) can bring up a new functioning instance and 2) updating a running system when new correct metadata exists.
I think that we'd probably end up spending a lot of time and energy making a seamless system if 1 and 2 were mixed--so that's a vote from me maybe for exploring some kind of "safe-update" first. For solving just 2, I think a publicly held and modifiable GHA file would be great, but having a clever cron job like you had before or using the API like you have above would be fine as well (or a mix).

**Comment by abessiari** [comment]
@jamesaoverton 
cc @kltm 

So I have checked in tools/safa-update.py that uses github api instead of travis. We can chat tomorrow about it in our meeting at 9:00 am.

**Comment by abessiari** [comment]
@jamesaoverton 

Ok we are ready to merge. 
Once the merge happens I will continue testing using the  master repo. 
Thanks.

**Comment by jamesaoverton** [comment]
Sorry for the delay. Thanks for all your work on this!

**Comment by jamesaoverton** [comment]
Thanks for all your work on this!

**Comment by abessiari** [comment]
Thanks.

**Pull request merged**